mlflow/models/model.py

```diff
@@ -747,9 +763,36 @@ def log(
                     await_registration_for=await_registration_for,
                     local_model_path=local_path,
                 )
-        model_info = mlflow_model.get_model_info()
-        if registered_model is not None:
-            model_info.registered_model_version = registered_model.version
+            model_info = mlflow_model.get_model_info()
+            if registered_model is not None:
+                model_info.registered_model_version = registered_model.version
+
+            # validate input example works for serving when logging the model
+            serving_input = mlflow_model.get_serving_input(local_path)
+            if mlflow_model.signature is None and serving_input is None:
+                _logger.warning(
+                    "Input example should be provided to infer model signature if the model "
+                    "signature is not provided when logging the model."
+                )
+            if serving_input:
+                from mlflow.pyfunc.scoring_server.utils import validate_serving_input
+
+                try:
+                    validate_serving_input(model_info.model_uri, serving_input)
+                except Exception as e:
+                    _logger.warning(
+                        f"Failed to validate serving input example {serving_input}. "
+                        "Alternatively, you can avoid passing input example and pass model "
+                        "signature instead when logging the model. To ensure the input example "
+                        "is valid prior serving, please try calling "
+                        "`mlflow.pyfunc.scoring_server.utils.validate_serving_input` on the model "
```

Comment:
Can we expose `validate_serving_input` in the `mlflow.pyfunc` module to minimize the number of public modules?

Reply:
I think scoring_server is supported in mlflow-skinny, so we can't expose it directly inside mlflow.pyfunc

Reply:
Isn't `mlflow.pyfunc` supported in mlflow-skinny?

Reply:
Nope, it's not supported in mlflow-skinny.
I create an env and only installed mlflow-skinny:
```
>>> from mlflow.pyfunc.scoring_server import _parse_json_data
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/serena.ruan/Documents/repos/mlflow/mlflow/pyfunc/__init__.py", line 402, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

Reply:
Let me have a try :) Just tried that mlflow.pyfunc is not supported in mlflow-skinny (missing numpy) anyways

Reply:
Emm mlflow.pyfunc.scoring_server imports something from mlflow.pyfunc, so if we import validate_serving_input in mlflow.pyfunc then we'll introduce a circular import

Reply:
Is the circular import fixable?

Reply:
Maybe yes, we can drop `PyFuncModel` type hint and import load_model within methods to avoid importing them at top level. But I always feel like this way to avoid circular imports might cause issue at some point (for example if a new method in mlflow.pyfunc.__init__ uses the method that internally imports mlflow.pyfunc)

Reply:
Can we use a lazy import in `mlflow.pyfunc`?

```python
def validate_serving_input(...):
    from mlflow.pyfunc.scoring_server.utils import _validate_serving_input

    return validate_serving_input(...)
```

Reply:
Updated!